com.unity.template.renderstreaming-hd/Assets/Scripts/RenderStreaming.cs

```diff
@@ -211,23 +250,118 @@ void OnOffer(ISignaling signaling, DescData e)
             while (opLocalDesc.MoveNext())
             {
             }
+
             if (opLocalDesc.IsError)
             {
-                Debug.LogError($"Network Error: {opLocalDesc.Error}");
+                Debug.LogError($"Network Error: {opLocalDesc.Error.message}");
                 return;
             }
 
             signaling.SendAnswer(connectionId, desc);
         }
 
+        RTCPeerConnection PrepareNewPeerConnection(ISignaling signaling, string connectionId, bool isOffer)
+        {
+            if (m_mapConnectionIdAndPeer.TryGetValue(connectionId, out var peer))
+            {
+                peer.Close();
+            }
+
+            var pc = new RTCPeerConnection();
+            m_mapConnectionIdAndPeer[connectionId] = pc;
+
+            pc.OnDataChannel = new DelegateOnDataChannel(channel => { OnDataChannel(pc, channel); });
+            pc.SetConfiguration(ref m_conf);
+            pc.OnIceCandidate = new DelegateOnIceCandidate(candidate =>
+            {
+                signaling.SendCandidate(connectionId, candidate);
+            });
+            pc.OnIceConnectionChange = new DelegateOnIceConnectionChange(state =>
+            {
+                if(state == RTCIceConnectionState.Disconnected)
+                {
+                    pc.Close();
+                    m_mapConnectionIdAndPeer.Remove(connectionId);
+                }
+            });
+
+            pc.OnTrack = trackEvent => { m_receiveStream.AddTrack(trackEvent.Track); };
+
+            pc.OnNegotiationNeeded = () => StartCoroutine(OnNegotiationNeeded(signaling, connectionId, isOffer));
+            return pc;
+        }
+
+        IEnumerator OnNegotiationNeeded(ISignaling signaling, string connectionId, bool isOffer)
+        {
+            if (!isOffer || m_negotiationneededCounter > 0)
```

Comment:
Does this mean allowing the renegotiation process only once?

Reply:
You're right, I think the `negotiationneededcounter` no need, so I'll turn it off.

Reply:
`SignalingState` is a better option than it.
It has possibilities that this become a bug, so it should be added the warning log to notify it.

Reply:
I added check about signalingstate and notify log
https://github.com/Unity-Technologies/UnityRenderStreaming/pull/334/files#diff-5439ce0e72ed855888098798e551fe66R315